### ISSUE TITLE
Wierd PLE divergence?

### ISSUE THREAD

**Issue by ranjitjhala** [ucsd-progsys/liquidhaskell#1371]
On an example due to  @PiotrJander

https://gist.github.com/PiotrJander/14c94f4ba401d8af75841eb96afca12c

(related to #1192?)

I _think_ it may be tickled by the odd (non-structural) recursion inside the nnf conversion.

**Comment by ranjitjhala** [comment]
Here's a "minimal" variant:

```haskell
{-@ LIQUID "--reflection" @-}
{-@ LIQUID "--ple"        @-}

module Tutorial6 where

import Prelude hiding (lookup)

data Pred
  = Var Int
  | Not Pred
  | Or  Pred Pred
  | And Pred Pred

{-@ reflect lookup @-}
lookup :: [a] -> a -> Int -> a
lookup []     def _ = def
lookup (v:vs) def k
  | k == 0          = v
  | k <  0          = def
  | otherwise       = lookup vs def (k-1)

{-@ reflect eval @-}
eval :: [Bool] -> Pred -> Bool
eval s (Var i)   = lookup s False i
eval s (And f g) = (eval s f) && (eval s g)
eval s (Or  f g) = (eval s f) || (eval s g)
eval s (Not f)   = not (eval s f)

{-@ reflect nnf @-}
nnf :: Pred -> Pred
nnf (Var i)    = (Var i)
nnf (And p q)  = And (nnf p) (nnf q)
nnf (Or  p q)  = Or  (nnf p) (nnf q)
nnf (Not f)    = nnf' f

{-@ reflect nnf' @-}
nnf' :: Pred -> Pred
nnf' (Not g)   = nnf g
nnf' (Var i)   = Not (Var i)
nnf' (Or  p q) = And (nnf' p) (nnf' q)                -- OK
nnf' (And p q) = Or  (nnf' p) (nnf' q)                -- OK
-- nnf' (Or  p q) = And (nnf (Not p)) (nnf (Not q))   -- PLE-DIVERGES
-- nnf' (And p q) = Or  (nnf (Not p)) (nnf (Not q))   -- PLE-DIVERGES

-- ODDLY, you need BOTH the PLE-DIVERGES cases to tickle the divergence...

{-@ thm :: s:[Bool] -> p:Pred -> { eval s p = eval s (nnf p) } @-}
thm :: [Bool] -> Pred -> Bool
thm s (Var i)         = True
-- thm s (And p q)       = thm s p && thm s q
-- thm s (Or  p q)       = thm s p && thm s q
-- thm s (Not p)         = undefined
thm _ _ = undefined
```

@atondwal if you get a chance, can you switch on the debug flags 
in `fixpoint` to see why PLE is diverging here, if you use 
the `-- PLE-DIVERGES` lines instead of the `OK` ones 
of course...

**Comment by ranjitjhala** [comment]
Here's a simpler function to demonstrate the same issue (I think):

```haskell
{-@ LIQUID "--reflection" @-}
{-@ LIQUID "--ple"        @-}

data Foo = Tick Foo | Tock Foo | Fin deriving (Show)

{-@ reflect tx @-}
tx :: Foo -> Foo
tx (Tock x) = Tock (tx x)
tx (Fin)    = Fin
tx (Tick x) = go x

{-@ reflect go @-}
go :: Foo -> Foo
go Fin      = Tick Fin
go (Tick x) = tx x
go (Tock x) = Tock (tx (Tick x))

{-@ thm :: x:Foo -> {x = tx x} @-}
thm :: Foo -> ()
thm Fin    = ()
thm (Tick x) = thm x
thm (Tock x) = thm x
```

**Comment by ranjitjhala** [comment]
Here's another occurrence of PLE diverging (TBD if its the _same_ bug or not) -- via @mossaka -- and wonder if the issues found by @rosekunkel are also related.

```haskell
{-@ LIQUID "--reflection"  @-}
{-@ LIQUID "--ple"         @-}

module Bug where

data Thing
  = N Int
  | Op Thing Thing

{-@ reflect killer @-}

{-@ killer :: _ -> _ -> { 1 + 2 == 3 } @-}
killer :: Thing -> Thing -> Thing
killer (N n) (Op e1 e2) = Op e1 (killer (N n) e2)
killer (Op e1 e2) x     = Op e1 (killer e2 x)
```

**Comment by ranjitjhala** [comment]
The above issues (PLE divergences) have been fixed by #1411 as described above, I've now added tests that check this in `tests/ple/{pos, neg}`.
